services/w3c/w3c-validation.service.js

```diff
@@ -0,0 +1,113 @@
+'use strict'
+const Joi = require('@hapi/joi')
+const { optionalUrl } = require('../validators')
+const {
+  documentation,
+  presetRegex,
+  getColor,
+  getMessage,
+  getSchema,
+} = require('./w3c-validation-helper')
+const { BaseJsonService } = require('..')
+
+const schema = Joi.object({
+  messages: Joi.array()
+    .required()
+    .items(
+      Joi.object({
+        type: Joi.string()
+          .allow('info', 'error', 'non-document-error')
```

Comment:
I was poking around the review app for this PR (https://shields-staging-pr-4148.herokuapp.com/) and tried testing with a non-existent URL for `targetUrl`. It looks like the response we get from the W3C service for the 'not found' scenario is an HTTP 200 with a single `non-document-error` error that has a message like `"message": "fakefakefakefakefakefakefake.com: Name or service not known"`. 

Internally Shields has defaults that help service classes deal with the 'not found' scenarios, but that's driven off the HTTP response codes (for example if an usptream API returns an HTTP 404, the default badge is something like ![](https://img.shields.io/badge/foo-not%20found-red)) so we'll have to handle this scenario explicitly here.

I'd suggest augmenting the `transform` function to check for this scenario (looks like we can check for a single error with a message that ends with `Name or service not known`), and to then throw Shields' `NotFound` error. 

For example:
```js
const { NotFound } = require('..')
...
  transform(messages) {
    ... // new logic to detect not found scenario
    if (notFound) {
      throw new NotFound({  prettyMessage: 'target url not found' })
    }
    return messages.reduce((accumulator, message) => {
....

```

Reply:
I tried implementing this but there is a false positive error from the validator. It handles the use case well when you pass it a URL that does not exist. But I found that it returns the same error message `{"url":"http://www.shields.io","messages":[{"type":"non-document-error","subType":"io","message":"c.validator.nu: Name or service not known"}]}` in one use case. The use case is when you pass the validator gets an url that exists but does not contain an iframe in the document and the schema for `HTML 4.01 Frameset, URL / XHTML 1.0 Frameset, URL`. Not sure if that is a bug or in the context of the validator it might not be wrong. Since it was trying to find an iframe which does not exists. I can change the service test to cover that false positive if that is the intention. To test the use case please run the following 3 calls and you can see the response.

1. https://validator.w3.org/nu/?doc=http://www.shields.io&out=json&schema=http://s.validator.nu/xhtml10/xhtml-frameset.rnc%20http://s.validator.nu/html4/assertions.sch%20http://c.validator.nu/all-html4/ (with frameset schemas that return an error of type 'non-document-error')
2. https://validator.w3.org/nu/?doc=http://www.shields.io&out=json (no schema and returns normal validation errors and warnings)

Reply:
deleted last message. I was wrong about the output

Reply:
Made the change to throw NotFound

Reply:
Thanks! Are you sure the check you've added fully covers the not found scenario? I'm still not seeing the `target url not found` message:
https://shields-staging-pr-4148.herokuapp.com/w3c-validation/default?targetUrl=https%3A%2F%2Fadfasdfasdfasdfadfadfadfasdfadf.com
![](https://shields-staging-pr-4148.herokuapp.com/w3c-validation/default?targetUrl=https%3A%2F%2Fadfasdfasdfasdfadfadfadfasdfadf.com)

and the json response from the w3c service does not seem to mention the 404 text in the error:
```json
{
  "url": "https://adfasdfasdfasdfadfadfadfasdfadf.com",
  "messages": [
     {
      "type": "non-document-error",
      "subType": "io",
      "message": "adfasdfasdfasdfadfadfadfasdfadf.com: Name or service not known"
    }
  ],
}
```

It'd also be good to add a service test to cover the not found scenario, since that's going to be entirely driven off the values in the API response from the upstream service

Reply:
Looks implement the NotFound is going to be tricky then. This validation `https://validator.w3.org/nu/?doc=http://hsivonen.com/test/moz/messages-types/warning.html&out=json&schema=http://s.validator.nu/xhtml10/xhtml-frameset.rnc%20http://s.validator.nu/html4/assertions.sch%20http://c.validator.nu/all-html4/` which checks an existing page with an info in normal HTML DTD returns the same error message as your test case. That is kind of a false positive.

Reply:
Since you changed the W3C checker from `https://validator.nu/` to `https://validator.w3.org/nu/`, are there any updates needed to the values we're providing for the schema? I ask because that error message is complaining about a value we are providing to the upstream API, not the user provided value for the target url

`"message": "c.validator.nu: Name or service not known"`

Reply:
There was no difference from what I tested so far. The same message appears on the validator.nu if I don't remember wrongly. There is a workaround I think. For the testcase you provided the hostname of the provided url is in the message instead of `c.validator.nu`. We could use that as a check

Reply:
Perhaps I'm doing something wrong, but I feel like I'm getting the expected response with https://validator.nu:

https://validator.nu/?doc=http://hsivonen.com/test/moz/messages-types/warning.html&out=json&schema=http://s.validator.nu/xhtml10/xhtml-frameset.rnc%20http://s.validator.nu/html4/assertions.sch%20http://c.validator.nu/all-html4/
```json
{ 
  "url":"http://hsivonen.com/test/moz/messages-types/warning.html",
  "messages":[ 
    { 
      "type":"error",
      "lastLine":6,
      "lastColumn":6,
      "firstColumn":1,
      "message":"Element “body” not allowed as child of element “html” in this context. (Suppressing further errors from this subtree.)",
      "extract":">\n</head>\n<body>\n<p cl",
      "hiliteStart":10,
      "hiliteLength":6
    },
    { 
      "type":"error",
      "lastLine":7,
      "lastColumn":13,
      "message":"“<” in an unquoted attribute value. Probable cause: Missing “>” immediately before.",
      "extract":"y>\n<p class=foo<bar></p>\n</bod",
      "hiliteStart":15,
      "hiliteLength":1
    },
    { 
      "type":"error",
      "lastLine":9,
      "lastColumn":7,
      "firstColumn":1,
      "message":"Element “html” is missing a required instance of child element “frameset”.",
      "extract":">\n</body>\n</html>",
      "hiliteStart":10,
      "hiliteLength":7
    }
  ]
}
```

but with `https://validator.w3.org/nu/` we're getting an error complaining about the values we are passing for the `schema` param, specifically for`c.validator.nu`:
https://validator.w3.org/nu/?doc=http://hsivonen.com/test/moz/messages-types/warning.html&out=json&schema=http://s.validator.nu/xhtml10/xhtml-frameset.rnc%20http://s.validator.nu/html4/assertions.sch%20http://c.validator.nu/all-html4/
``` json
{ 
  "url":"http://hsivonen.com/test/moz/messages-types/warning.html",
  "messages":[ 
    { 
      "type":"non-document-error",
      "subType":"io",
      "message":"c.validator.nu: Name or service not known"
    }
  ]
}
```

It seems like there's some correlation between the validator endpoint we are using and whether the schema values work

Reply:
Ok looks like the preset and schema might not be the best way forward. I am going to look into this more. Sorry for the troubles

Reply:
> Ok looks like the preset and schema might not be the best way forward

Well, you're definitely closer to the details on this, but my initial question is simply whether we'd need to use different schema _values_ with https://validator.w3.org/nu/ (maybe instead of  `c.validator.nu` we'd need something like `c.validator.w3...`)? 

It also seems like things are working as-is with this implementation with https://validator.nu, so before we go down an entirely different road I think it would be worth revisiting using `https://validator.nu/` vs. `https://validator.w3.org/nu/` as our upstream endpoint. FWIW, Istarted getting HTTP 429 errors with https://validator.w3.org/nu/ after only a handful of requests which has me worried about rate limiting with that endpoint. I've not seen any such issues with https://validator.nu

Reply:
It was interesting since `https://validator.nu/` or `https://validator.w3.org/nu/` appear to be running off the same source `https://github.com/validator/validator`.  https://about.validator.nu actually asks you to report issues to https://github.com/validator/validator. Given that you faced the limiting issue. I redirect the code to point back to https://validator.nu. I also found a new param doctype. This appears to be cleaner then using the schema which is what the `https://validator.nu/` UI is using. Although I tested against the API version of it `https://validator.nu/?doc=https%3A%2F%2Fhsivonen.com%2Ftest%2Fmoz%2Fmessages-types%2Fwarning.html&charset=%28detect+automatically%29&doctype=ISO%2FIEC+15445%3A2000+%28"ISO+HTML"%29&out=json` and it works. I am working if we should move of schema and use that instead.

Reply:
very interesting indeed. if we've got things working here, then i'd be inclined to finish this out with the current-implementation (plus support for 'not found' scenario). if you're willing and interested in investigating an alternative then we can tackle that in a follow up

Reply:
I think the last commit has the changes required for NoFound then. I added a new testcase for it as well. Thanks.

Reply:
https://shields-staging-pr-4148.herokuapp.com/w3c-validation/default?targetUrl=https%3A%2F%2Fgasdfasdfithub.com
![](https://shields-staging-pr-4148.herokuapp.com/w3c-validation/default?targetUrl=https%3A%2F%2Fgasdfasdfithub.com)
🙌